agent-codeless/src/main/java/com/microsoft/applicationinsights/agentc/internal/diagnostics/status/StatusFile.java

```diff
@@ -0,0 +1,197 @@
+
+        if (SystemInformation.INSTANCE.isWindows()) {
+            directory = "D:" + DEFAULT_STATUS_FILE_DIRECTORY;
+        } else {
+            directory = DEFAULT_STATUS_FILE_DIRECTORY;
+        }
+    }
+
+    public static <T> void putValueAndWrite(String key, T value) {
+        if (shouldNotWrite()) {
+            return;
+        }
+        CONSTANT_VALUES.put(key, value);
+        write();
+    }
+
+    public static void write() {
+        if (shouldNotWrite()) {
+            return;
+        }
+        WRITER_THREAD.submit(new Runnable() {
+            @Override
+            public void run() {
+                Map<String, Object> map = getJsonMap();
+
+                String fileName = constructFileName(map);
+
+                synchronized (StatusFile.class) { // the executor should prevent more than one thread from executing this block. this is just a safeguard
+                    final File file = new File(directory, fileName);
+                    boolean dirsWereCreated = file.getParentFile().mkdirs();
+                    if (dirsWereCreated || file.getParentFile().exists()) {
+                        BufferedSink b = null;
+                        try {
+                            b = getBuffer(file);
+                            new Builder().build().adapter(Map.class).indent(" ").nullSafe().toJson(b, map);
+                            b.flush();
+                        } catch (Exception e) {
+                            LoggerFactory.getLogger(StatusFile.class).error("Error writing {}", file.getAbsolutePath(), e);
+                            if (b != null) {
+                                try {
+                                    b.close();
+                                } catch (IOException ex) {
+                                    // ignore this
+                                }
+                            }
+                        }
+                    } else {
+                        LoggerFactory.getLogger(StatusFile.class).error("Parent directories for status file could not be created: {}", file.getAbsolutePath());
+                    }
+                }
+            }
+        }, "StatusFileJsonWrite");
+    }
+
+    private static synchronized BufferedSink getBuffer(File file) throws IOException {
+        if (buffer != null) {
+            buffer.close();
+        }
+        buffer = Okio.buffer(Okio.sink(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.DELETE_ON_CLOSE, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING));
+        return buffer;
+    }
+
+    @VisibleForTesting
+    static Map<String, Object> getJsonMap() {
+        Map<String, Object> map = new LinkedHashMap<>(CONSTANT_VALUES);
+        for (DiagnosticsValueFinder finder : VALUE_FINDERS) {
+            final String value = finder.getValue();
+            if (!Strings.isNullOrEmpty(value)) {
+                map.put(capitalize(finder.getName()), value);
```

Comment:
Is `capitalize()` needed? it looks like names are already capitalized?

Reply:
Not all are. I'm keeping the same style as existing status files.